### ISSUE TITLE
double free or corruption when stopping

### ISSUE THREAD

**Issue by nerijus** [Tarsnap/spiped#201]
Version 1.6.0 on CentOS 7, when I do systemctl restart spiped@remote1-sentinel.service, I get:
```
Rgp 01 15:52:20 host systemd[1]: Stopping Spipe tunnel daemon service remote1-sentinel...
Rgp 01 15:52:21 host spiped[19882]: *** Error in `/usr/local/bin/spiped': double free or corruption (!prev): 0x00000000013597a0 ***
Rgp 01 15:52:21 host spiped[19882]: ======= Backtrace: =========
Rgp 01 15:52:21 host spiped[19882]: /lib64/libc.so.6(+0x81609)[0x7f5f4ad91609]
Rgp 01 15:52:21 host spiped[19882]: /usr/local/bin/spiped[0x408d22]
Rgp 01 15:52:21 host spiped[19882]: /usr/local/bin/spiped[0x4092b9]
Rgp 01 15:52:21 host spiped[19882]: /usr/local/bin/spiped[0x402f9b]
Rgp 01 15:52:21 host spiped[19882]: /lib64/libc.so.6(__libc_start_main+0xf5)[0x7f5f4ad32495]
Rgp 01 15:52:21 host spiped[19882]: /usr/local/bin/spiped[0x403013]
Rgp 01 15:52:21 host spiped[19882]: ======= Memory map: ========
Rgp 01 15:52:21 host spiped[19882]: 00400000-00412000 r-xp 00000000 fd:00 25524780                           /usr/local/bin/spiped
Rgp 01 15:52:21 host spiped[19882]: 00611000-00612000 r--p 00011000 fd:00 25524780                           /usr/local/bin/spiped
Rgp 01 15:52:21 host spiped[19882]: 00612000-00613000 rw-p 00012000 fd:00 25524780                           /usr/local/bin/spiped
Rgp 01 15:52:21 host spiped[19882]: 01357000-01378000 rw-p 00000000 00:00 0                                  [heap]
Rgp 01 15:52:21 host spiped[19882]: 7f5f44000000-7f5f44021000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f44021000-7f5f48000000 ---p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f498ac000-7f5f498c1000 r-xp 00000000 fd:00 25838600                   /usr/lib64/libgcc_s-4.8.5-20150702.so.1
Rgp 01 15:52:21 host spiped[19882]: 7f5f498c1000-7f5f49ac0000 ---p 00015000 fd:00 25838600                   /usr/lib64/libgcc_s-4.8.5-20150702.so.1
Rgp 01 15:52:21 host spiped[19882]: 7f5f49ac0000-7f5f49ac1000 r--p 00014000 fd:00 25838600                   /usr/lib64/libgcc_s-4.8.5-20150702.so.1
Rgp 01 15:52:21 host spiped[19882]: 7f5f49ac1000-7f5f49ac2000 rw-p 00015000 fd:00 25838600                   /usr/lib64/libgcc_s-4.8.5-20150702.so.1
Rgp 01 15:52:21 host spiped[19882]: 7f5f49ac2000-7f5f49ac3000 ---p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f49ac3000-7f5f4a2c3000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a2c3000-7f5f4a2d9000 r-xp 00000000 fd:00 25838616                   /usr/lib64/libresolv-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a2d9000-7f5f4a4d8000 ---p 00016000 fd:00 25838616                   /usr/lib64/libresolv-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a4d8000-7f5f4a4d9000 r--p 00015000 fd:00 25838616                   /usr/lib64/libresolv-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a4d9000-7f5f4a4da000 rw-p 00016000 fd:00 25838616                   /usr/lib64/libresolv-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a4da000-7f5f4a4dc000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a4dc000-7f5f4a4e1000 r-xp 00000000 fd:00 25650242                   /usr/lib64/libnss_dns-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a4e1000-7f5f4a6e1000 ---p 00005000 fd:00 25650242                   /usr/lib64/libnss_dns-2.17.so
Rgp 01 15:52:21 host systemd[1]: spiped@remote1-sentinel.service: main process exited, code=killed, status=6/ABRT
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a6e1000-7f5f4a6e2000 r--p 00005000 fd:00 25650242                   /usr/lib64/libnss_dns-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a6e2000-7f5f4a6e3000 rw-p 00006000 fd:00 25650242                   /usr/lib64/libnss_dns-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a6e3000-7f5f4a6ef000 r-xp 00000000 fd:00 25838603                   /usr/lib64/libnss_files-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a6ef000-7f5f4a8ee000 ---p 0000c000 fd:00 25838603                   /usr/lib64/libnss_files-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a8ee000-7f5f4a8ef000 r--p 0000b000 fd:00 25838603                   /usr/lib64/libnss_files-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a8ef000-7f5f4a8f0000 rw-p 0000c000 fd:00 25838603                   /usr/lib64/libnss_files-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a8f0000-7f5f4a8f6000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a8f6000-7f5f4a90b000 r-xp 00000000 fd:00 25838638                   /usr/lib64/libz.so.1.2.7
Rgp 01 15:52:21 host spiped[19882]: 7f5f4a90b000-7f5f4ab0a000 ---p 00015000 fd:00 25838638                   /usr/lib64/libz.so.1.2.7
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ab0a000-7f5f4ab0b000 r--p 00014000 fd:00 25838638                   /usr/lib64/libz.so.1.2.7
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ab0b000-7f5f4ab0c000 rw-p 00015000 fd:00 25838638                   /usr/lib64/libz.so.1.2.7
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ab0c000-7f5f4ab0e000 r-xp 00000000 fd:00 25524792                   /usr/lib64/libdl-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ab0e000-7f5f4ad0e000 ---p 00002000 fd:00 25524792                   /usr/lib64/libdl-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ad0e000-7f5f4ad0f000 r--p 00002000 fd:00 25524792                   /usr/lib64/libdl-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ad0f000-7f5f4ad10000 rw-p 00003000 fd:00 25524792                   /usr/lib64/libdl-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4ad10000-7f5f4aed2000 r-xp 00000000 fd:00 25279505                   /usr/lib64/libc-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4aed2000-7f5f4b0d2000 ---p 001c2000 fd:00 25279505                   /usr/lib64/libc-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b0d2000-7f5f4b0d6000 r--p 001c2000 fd:00 25279505                   /usr/lib64/libc-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b0d6000-7f5f4b0d8000 rw-p 001c6000 fd:00 25279505                   /usr/lib64/libc-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b0d8000-7f5f4b0dd000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b0dd000-7f5f4b0e4000 r-xp 00000000 fd:00 25838618                   /usr/lib64/librt-2.17.so
Rgp 01 15:52:21 host systemd[1]: Stopped Spipe tunnel daemon service remote1-sentinel.
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b0e4000-7f5f4b2e3000 ---p 00007000 fd:00 25838618                   /usr/lib64/librt-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b2e3000-7f5f4b2e4000 r--p 00006000 fd:00 25838618                   /usr/lib64/librt-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b2e4000-7f5f4b2e5000 rw-p 00007000 fd:00 25838618                   /usr/lib64/librt-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b2e5000-7f5f4b2fc000 r-xp 00000000 fd:00 25838612                   /usr/lib64/libpthread-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b2fc000-7f5f4b4fb000 ---p 00017000 fd:00 25838612                   /usr/lib64/libpthread-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b4fb000-7f5f4b4fc000 r--p 00016000 fd:00 25838612                   /usr/lib64/libpthread-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b4fc000-7f5f4b4fd000 rw-p 00017000 fd:00 25838612                   /usr/lib64/libpthread-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b4fd000-7f5f4b501000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b501000-7f5f4b736000 r-xp 00000000 fd:00 25386506                   /usr/lib64/libcrypto.so.1.0.2k
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b736000-7f5f4b936000 ---p 00235000 fd:00 25386506                   /usr/lib64/libcrypto.so.1.0.2k
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b936000-7f5f4b952000 r--p 00235000 fd:00 25386506                   /usr/lib64/libcrypto.so.1.0.2k
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b952000-7f5f4b95f000 rw-p 00251000 fd:00 25386506                   /usr/lib64/libcrypto.so.1.0.2k
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b95f000-7f5f4b963000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4b963000-7f5f4b985000 r-xp 00000000 fd:00 25166930                   /usr/lib64/ld-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4bb76000-7f5f4bb7b000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4bb82000-7f5f4bb84000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7f5f4bb84000-7f5f4bb85000 r--p 00021000 fd:00 25166930                   /usr/lib64/ld-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4bb85000-7f5f4bb86000 rw-p 00022000 fd:00 25166930                   /usr/lib64/ld-2.17.so
Rgp 01 15:52:21 host spiped[19882]: 7f5f4bb86000-7f5f4bb87000 rw-p 00000000 00:00 0
Rgp 01 15:52:21 host spiped[19882]: 7ffd71903000-7ffd71924000 rw-p 00000000 00:00 0                          [stack]
Rgp 01 15:52:21 host spiped[19882]: 7ffd71938000-7ffd7193a000 r-xp 00000000 00:00 0                          [vdso]
Rgp 01 15:52:21 host spiped[19882]: ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
Rgp 01 15:52:21 host systemd[1]: Unit spiped@remote1-sentinel.service entered failed state.

```
spiped@remote1-sentinel.service is a link to /etc/systemd/system/spiped@.service:
```
[Unit]
Description=Spipe tunnel daemon service %i
After=network-online.target

[Service]
Type=simple
EnvironmentFile=/etc/spiped/%i
ExecStart=/usr/local/bin/spiped -${mode} -s ${source} -t ${target} -k ${key} -F -g
RestartSec=1
Restart=always
StartLimitInterval=0

[Install]
WantedBy=multi-user.target
```

**Comment by nerijus** [comment]
It does not happen with an older version from a few years ago.

**Comment by gperciva** [comment]
Hmm, I'm trying to reproduce this with CentOS 7 inside virtualbox.

- is it possible to reproduce this on the command-line, without involving `systemd`?
- what do you have in `/etc/spiped/` ?
- I see that you're using `/usr/local/bin/spiped`.  Did you compile it with the default `make`, or `clang`, or...?

**Comment by nerijus** [comment]
I cannot reproduce it without systemd. systemd starts it like this:
```
/usr/local/bin/spiped -e -s 127.0.0.1:26390 -t host:36390 -k /etc/spiped/default.key -F -g
```
If I run this command manually, and press Ctrl-C, it exits cleanly without any output.

I compiled it by running make.

**Comment by cperciva** [comment]
Might be related to the "graceful shutdown" code?  That code wouldn't run if you just ^C.

**Comment by nerijus** [comment]
I was able to get coredump without systemd by killing and executing spiped (after a few tries):
```
# gdb ./spiped /tmp/core_spiped.9027 
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-114.el7
...
Reading symbols from /tmp/spiped-1.6.0/spiped/spiped...done.
[New LWP 9027]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/tmp/spiped-1.6.0/spiped/spiped -e -s 127.0.0.1:26390 -t host:3639'.
Program terminated with signal 6, Aborted.
#0  0x00007f3aa4d812c7 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
55	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) bt
#0  0x00007f3aa4d812c7 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007f3aa4d829b8 in __GI_abort () at abort.c:90
#2  0x00007f3aa4dc3e17 in __libc_message (do_abort=do_abort@entry=2, fmt=fmt@entry=0x7f3aa4ed57f8 "*** Error in `%s': %s: 0x%s ***\n")
    at ../sysdeps/unix/sysv/linux/libc_fatal.c:196
#3  0x00007f3aa4dcc609 in malloc_printerr (ar_ptr=0x7f3aa5111760 <main_arena>, ptr=<optimized out>, str=0x7f3aa4ed5920 "double free or corruption (!prev)", action=3)
    at malloc.c:5004
#4  _int_free (av=0x7f3aa5111760 <main_arena>, p=<optimized out>, have_lock=0) at malloc.c:3843
#5  0x000000000040d983 in sock_addr_free (sa=0x1fd4ce0) at ../libcperciva/util/sock.c:448
#6  0x000000000040d9bf in sock_addr_freelist (sas=0x1fd2a40) at ../libcperciva/util/sock.c:467
#7  0x00000000004038d9 in main (argc=0, argv=0x7ffdbd676d78) at main.c:344
```
spiped was recompiled by changing `CFLAGS_DEFAULT= -g` in Makefile.

**Comment by gperciva** [comment]
Thanks, that narrows it down considerably!

**Comment by gperciva** [comment]
I think I found it, although on my system it shows up as segmentation fault rather than a double free.

I believe that this was fixed by Michael Weiser in:
https://github.com/Tarsnap/spiped/commit/8faf99d81bd9b9d8e84f47920f21559642b065cd
https://github.com/Tarsnap/spiped/pull/153

To reproduce the problem in 1.6.0, run:
```
./spiped/spiped -e -s 0.0.0.0:8082 -t mac:8081 -k k.txt -r 1 -F
```

The `-r 1` is important -- this re-resolves the target address every 1 seconds.  The default is 60 seconds, so if you try to manually force the crash you'd need to wait a minute before killing the process.  (That's why I didn't manage to reproduce it until now.)

Stopping the process with `^C` does not cause a segfault, but using `kill` does produce one.

Running that same command in git `master` does not produce a fault.  I also tested it in a new branch, `fix-1.6.0-sigsegv`, and it did not have a problem.  I've uploaded that branch in case you're interested in testing it.
